Add KCL, KFramework Languages

mihaigalos opened the pull request:
<!--- Briefly describe your changes in the field above. -->

## Description
<!--- If necessary, go into depth of what this pull request is doing. -->
This PR was initially produced to offer support for the `KCL` language. During review, we found out that the `*.k` file extension may conflict with other languages so the PR addresses both `KCL` and `KFramework` languages.

## Checklist:


- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      -  https://github.com/search?type=code&q=NOT+is%3Afork+path%3A*.k+kcl
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - https://github.com/kcl-lang/examples
    - Sample license(s): Apache - https://www.kcl-lang.io/docs/community/intro/license
  - [x] I have included a syntax highlighting grammar: https://github.com/kcl-lang/vscode-kcl.git
      <!-- Setting a color is strongly recommended, but optional: `#cccccc` is used by default -->
  - [x] I have added a color
    - Hex value: `##00ADD8`
    - Rationale: This color represents one of the 4 colours of the [KCL logo](https://www.kcl-lang.io/). <!-- Please specify why you chose this color (if it was randomly selected, please say so); it helps arbitrate future requests to change a language's color -->
  - [x] I have updated the heuristics to distinguish my language from others using the same extension.
  
  
- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      -  https://github.com/search?type=code&q=NOT+is%3Afork+path%3A*.k
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - https://github.com/runtimeverification/k/tree/master/pyk/regression-new
    - Sample license(s): BSD 3 Clause - https://github.com/runtimeverification/k/blob/master/LICENSE.md
  - [x] I have included a syntax highlighting grammar: https://github.com/LucianCumpata/K-VSCode
  <!-- Setting a color is strongly recommended, but optional: `#cccccc` is used by default -->
  - [x] I have added a color
    - Hex value: `#4195c5`
    - Rationale: This color represents the color used in the  [K Framework](https://www.kcl-lang.io/) logo. <!-- Please specify why you chose this color (if it was randomly selected, please say so); it helps arbitrate future requests to change a language's color -->
  - [x] I have updated the heuristics to distinguish my language from others using the same extension.

alexdaima commented:
@mihaigalos

Me and my team are heavy users of KCL, I am willing to provide any assistance for this feature if needed :)

lildude commented on lib/linguist/languages.yml:
```diff
@@ -3725,6 +3725,16 @@ Just:
+  type: programming
+  color: "#00ADD8"
+  tm_scope: source.kcl
+  filenames:
+  - ".k"
```
This is for specific files which generally don't have an extension that matches that being added.

```suggestion
```

mihaigalos replied:
thx. I'll do a bigger refactoring.

lildude reviewed (changes requested):
The `.k` looks to be incredibly popular and most of those files don't look like KCL from what I can determine from your samples. 

If we were to merge this PR now, all of those files will be incorrectly classified as KCL which we don't want to do.

In order for this PR to be merged, you will need to identify at least one other language that is using the `.k` extension, ideally the most popular language, and add support for it as part of this PR.

alexdaima commented:
@lildude @mihaigalos 

I can confirm the samples provided with the `.k` extension in this PR are very much KCL, a few references below show some of the documentation for the samples that were provided:

- [Schema](https://www.kcl-lang.io/docs/reference/lang/spec/schema)
- [Lambda Functions](https://www.kcl-lang.io/docs/reference/lang/tour#function)
- [YAML Streams](https://www.kcl-lang.io/docs/reference/model/manifests#yaml_stream)

As for the property `filenames:` I agree that while `.k` **should** be removed, there should actually be the addition of `kcl.mod` and `kcl.mod.lock`, both of which are always called as such ([KCL Package Manifest file](https://www.kcl-lang.io/docs/user_docs/guides/package-management/how-to/kcl_mod)) and are used heavily when installing/importing packages/modules from [ArtifactHub](https://artifacthub.io/packages/search?repo=kcl-module&sort=relevance&page=1).

Note that actually in Github `kcl.mod` files ([Examples](https://github.com/search?q=repo%3Akcl-lang%2Fexamples++language%3AAMPL&type=code)) are currently being incorrectly interpreted as [AMPL](https://github.com/ampl).

Here are a few other resources that match against the `.k` extension in order to associate with KCL:

- [KCL VSCode Extension](https://github.com/kcl-lang/vscode-kcl/blob/a3cf72316d14775f8fc9af1fd1eabf684332e0bf/package.json#L70)
- [KCL NeoVim Extension](https://github.com/kcl-lang/kcl.nvim/blob/beededb5a8ed01ba2d121ed053aa6380ecb59597/ftdetect/kcl.lua#L3)
- [KCL Zed Extension](https://github.com/pisarenko91/kcl-zed-extension/blob/9751413a52e1e8edd5c129ed1517de13c9f7f83d/languages/kcl/config.toml#L3)
- [KCL IntelliJ Extension](https://github.com/kcl-lang/intellij-kcl/blob/11cc8d57a6e0a20cec4c52de4e0cc57812988d48/src/main/java/io/kusionstack/kcl/KCLFileType.java#L37)

Other languages that seem to use `.k`:

- [K Framework](https://github.com/runtimeverification/k) does use `.k` files and is in active usage, but with a very different syntax. An example can be found [here](https://github.com/runtimeverification/k/blob/master/k-distribution/tests/regression-new/concrete-function/a1-spec.k).
- Was once used by [K Language](https://en.wikipedia.org/wiki/K_(programming_language)), which was commercialized by [KX Systems](https://en.wikipedia.org/wiki/KX_Systems), but the K language now just stands as a predecessor of the [Q Language](https://en.wikipedia.org/wiki/Q_(programming_language_from_Kx_Systems)) which is used in their kdb+ database.

mihaigalos commented:
Hi @lildude, @alexdaima,

I've addressed your comments and added the KFramework as part of this PR.
The jobs are failing because the license for [K-VSCode.dev.yml](https://github.com/github-linguist/linguist/pull/7490/files#diff-3e98d25cc5223d918fc0ff1451faf219d2a01da9f5334a5c82328190f352aa6d) is not yet allowed. Can you please have a look?

alexdaima commented on vendor/licenses/git_submodule/K-VSCode.dep.yml:
```diff
@@ -0,0 +1,31 @@
+version: c6d017b7225d3877e60edb4af5463cc624cd6d7d
+type: git_submodule
+homepage: https://github.com/LucianCumpata/K-VSCode.git
+license: other
```
Possibly because this needs to be `mit`?

mihaigalos replied:
I don't think so. This file is auto-generated and I presume GH needs to approve the ones that are set to "other" - there are examples in that folder.

mihaigalos commented:
Hi @lildude - can you please have a look again? Thanks.

lildude commented on vendor/licenses/git_submodule/K-VSCode.dep.yml:
```diff
@@ -0,0 +1,31 @@
+version: c6d017b7225d3877e60edb4af5463cc624cd6d7d
+type: git_submodule
+homepage: https://github.com/LucianCumpata/K-VSCode.git
+license: other
```
Can override this because the auto-detection hasn't been able to determine it.

```suggestion
license: mit
```

alexdaima commented:
@lildude 

As checks are now passing, is the PR ready to be merged and included now?

lildude reviewed (approved):
LGTM. Thanks.

> [!IMPORTANT]
> The changes in this PR will not appear on GitHub until the next release has been made and deployed. See [here](https://github.com/github/linguist/blob/master/docs/troubleshooting.md#my-linguist-pr-has-been-merged-but-github-doesnt-reflect-my-changes) for more details.
